question-servers/calculation.js

```diff
@@ -1,221 +1,195 @@
+      message: error.message,
+      stack: error.stack,
     });
-  },
-};
+  }
+
+  return JSON.stringify(result);
+}
+
+/**
+ * Runs both the `control` and `candidate` functions at the same time and
+ * compares their output. No matter what the candidate does, the results of
+ * the `control` function will always be what's returned to the client.
+ *
+ * @param {string} name
+ * @param {Function} control
+ * @param {Function} candidate
+ * @returns {Function}
+ */
+function questionFunctionExperiment(name, control, candidate) {
+  const experiment = experimentAsync({
+    name,
+    control: promisifyQuestionFunction('control', control),
+    candidate: promisifyQuestionFunction('candidate', candidate),
+    options: {
+      publish: ({ controlResult, controlError, candidateResult, candidateError }) => {
+        // Grab the current span from context - this is the span created with
+        // `startActiveSpan` below.
+        const span = trace.getSpan(context.active());
+
+        // The control implementation does not utilize course issues.
+        const controlHasError = !!controlError;
+
+        // The candidate implementation does propagate some errors as course
+        // issues, so we need to take those into account when checking if the
+        // implementation resulted in an error.
+        const candidateHasError = !!candidateError || !_.isEmpty(candidateResult?.[0]);
+
+        // We don't want to assert that the errors themselves are equal, just
+        // that either both errored or both did not error.
+        const errorsMismatched = controlHasError !== candidateHasError;
+
+        // The "results" can actually contain error information for the candidate,
+        // but not for the control. To avoid false positives, we'll only compare
+        // the "data" portion of the results. The "course issues" portion of the
+        // results was already handled above.
+        const controlData = controlResult?.[1];
+        const candidateData = candidateResult?.[1];
+        const controlHasData = !_.isEmpty(controlData);
+        const candidateHasData = !_.isEmpty(candidateData);
+
+        // Lodash is clever enough to understand Buffers, so we don't event need to
+        // special-case `getFile`, which can sometimes return a Buffer.
+        //
+        // We only assert that `controlData` and `candidateData` are equal if
+        // they're both not "empty" (empty object, null, undefined, etc.) since
+        // for our purposes, all empty data is equal.
+        const dataMismatched =
+          controlHasData && candidateHasData && !_.isEqual(controlData, candidateData);
+
+        if (errorsMismatched || dataMismatched) {
+          span.setAttributes({
+            'experiment.result': 'mismatched',
+            'experiment.control': observationPayload(controlError, controlResult),
+            'experiment.candidate': observationPayload(candidateError, candidateResult),
+          });
+          span.setStatus({
```

Comment:
@mwest1066 now that Sentry is a part of our stack, I'm wondering if perhaps we should use that instead to report any mismatches instead of Honeycomb? Its lossless retention of errors seems to be what we want, without any need for messing around with sample rates.

Reply:
Agreed! I think that Sentry is just a better fit for error reporting than HC.

Reply:
I made the switch in 00aaba2.